### PR TITLE
Add `prefer-native-coercion-functions` rule

### LINKED ISSUE #1564: Rule proposal: `prefer-filter-boolean`
It's more readable that returned values will be only `truthy`

## Fail

```js
list.filter(v => v)
list.filter(v => !!v)
list.filter(v => Boolean(v))
```

## Pass

```js
list.filter(Boolean)
list.filter(v => !v)
list.filter(v => v === true)
list.filter(v => v > 0)
```

### LINKED ISSUE #1643: Rule proposal: `prefer-native-coercion-functions`
### Description

I want to propose a more general rule to replace #1564.

[Primitive wrapper objects in JavaScript](https://developer.mozilla.org/en-US/docs/Glossary/Primitive)

When function is equivalence to the primitive wrapper, it should be used directly, including `String`, `Number`, `BigInt`, `Boolean`, and `Symbol `.

### Fail

```js
const toString = (v) => String(v);
```

```js
const toString = (v) => `${v}`;
```

```js
const toString = (v) => '' + v;
```

```js
const toNumber = (v) => Number(v);
```

```js
const toNumber = (v) => +v;
```

```js
const toBoolean = (v) => Boolean(v);
```

```js
const toBoolean = (v) => !!v;
```

```js
// These functions don't have to be arrow function or function expression/declaration
useFunction(function (v) {
	return Boolean(v);
});
```

```js
/* 
These callbacks not really equivalence to `Boolean`, but the result is used as `Boolean`,
Addtional check to cover cases in #1564
*/
list.every(v => v);
list.filter(v => v);
list.some(v => v);
list.find(v => v);
list.findIndex(v => v);
```

### Pass

```js
const toString = String;
```

```js
useFunction(Boolean);
```

```js
list.every(Boolean);
list.filter(Boolean);
list.some(Boolean);
list.find(Boolean);
list.findIndex(Boolean);
```

Some cases already coverd by builtin rule [no-implicit-coercion](https://eslint.org/docs/rules/no-implicit-coercion), so maybe we can assume some type-casting is already in shape of `v => Boolean(v)`

### PR DESCRIPTION
**Pull request by fisker**
Fixes #1643
Fixes #1564

### CODE REVIEW

File: rules/prefer-native-coercion-functions.js [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+			node,
+			loc: getFunctionHeadLocation(node, sourceCode),
+			messageId: MESSAGE_ID,
+			data: {
+				functionNameWithKind: getFunctionNameWithKind(node, sourceCode),
+				replacementFunction,
+			},
+		};
+
+		// Do not fix, if there are comments or extra parameters
```
[the remark this answers is not in the document]
Comment (sindresorhus) [reply rules/prefer-native-coercion-functions.js:156]: A short sentence on why we don't fix here would be useful to future readers of the code.

**Comment by fisker** [comment]
I didn't notice this is still a draft. 😄

File: rules/prefer-native-coercion-functions.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (fisker) [reply rules/prefer-native-coercion-functions.js]: Explain why we should not remove comments? Isn't it obvious?

[the remark this answers is not in the document]
Comment (fisker) [reply rules/prefer-native-coercion-functions.js]: https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1767/commits/0193d5ca1d5c2a65849bf01318d83b0c102f5333

**Pull request merged**